File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root==null)
            return new ArrayList<List<Integer>>() {{}};
        //to track from which side we have to print we keeps a flag variable
        //it is a level order traversal 
        Queue<TreeNode> q=new LinkedList<>();
        //adding root node
        q.add(root);
        boolean flag=true;//true reperesent from left to right;
        List<List<Integer>> ans=new ArrayList<List<Integer>>();
        while(!q.isEmpty()){
            
            int size=q.size();
            List<Integer> tempAns=new ArrayList<>();
            for(int i=0;i<size;i++){
                TreeNode frontNode=q.poll();
                if(flag)
                tempAns.add(frontNode.val);
                else
                   tempAns.add(0,frontNode.val); 
                //now push left and right child of root node
                if(frontNode.left!=null)
                    q.add(frontNode.left);
                if(frontNode.right!=null)
                    q.add(frontNode.right);
            }
            //reverse flag direction
            flag=!flag;
           ans.add(tempAns);
            
        }
        return ans;
    }
}

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(root==NULL)
            return ans;
        
        queue<TreeNode*> q;
        q.push(root);
        bool flag=false;//false represents from left to right
        while(!q.empty()){
            int s=q.size();
            vector<int> temp;
            for(int i=0;i<s;i++){
                TreeNode* fnt=q.front();
                q.pop();
                // temp.push_back(fnt->val);
                
//                 if(flag!=true){
                    
                
//                 if(fnt->left!=NULL)
//                     q.push(fnt->left);
//                 if(fnt->right!=NULL)
//                     q.push(fnt->right);
                
//                 }
//                 else{
                   
//                 if(fnt->right!=NULL)
//                     q.push(fnt->right);
//                  if(fnt->left!=NULL)
//                     q.push(fnt->left);
//                 }
                //so if we follow the above approach ie. if true then insert right node first in queue and left node thereafter, we got wrong answer in level 2 if right node have child also 
                // ex- [1,2,3,4,5,6,7] gives wrong ans
                //to correct above approach we can make changes in inserting ele into vector
                //so when we are inserting ele into temp vector we check flag if its true 
                // then insert from beg in vector
                
                if(flag!=true)  temp.push_back(fnt->val);
                else temp.insert(temp.begin(),fnt->val);
                 if(fnt->left!=NULL)
                    q.push(fnt->left);
                if(fnt->right!=NULL)
                    q.push(fnt->right);
                
            }
            flag=!flag;

            //push the temp into ans
            ans.push_back(temp);
            
        }
        return ans;
    }
};